netket/driver/qsr.py

```diff
@@ -0,0 +1,1010 @@
+
+            Exponentially expensive in the hilbert space size!
+
+            Need to know the target state!
+
+        """
+        if len(target_state.shape) == 1:
+            target_state = target_state.reshape(-1, 1)
+            target_state = target_state @ target_state.conj().T
+        if self._entropy is not None and not no_cache:
+            return self._entropy
+        rotations = self._training_rotations[::n_shots]
+        entropy_list = []
+        for rot in rotations:
+            rho_rot = (rot @ (rot @ target_state).conj().T).conj().T
+            prob = np.real(np.diag(rho_rot))
+            entropy = -np.sum(prob * np.log(prob + 1e-10))
+            entropy_list.append(entropy)
+        if no_cache:
+            return np.mean(entropy_list)
+        self._entropy = np.mean(entropy_list)
+        return self._entropy
+
+    def KL(self, target_state: Optional[Array] = None, n_shots: Optional[int] = None):
+        r"""
+        Compute average KL divergence loss over a batch of data.
+
+        Args:
+            target_state: the target state.
+            n_shots: number of shots per measurement basis.
+
+        .. warn::
+
+            Exponentially expensive in the hilbert space size!
+
+            Need to know the target state!
+
+        """
+        return statistics(
+            self.nll(return_stats=False) - self.entropy(target_state, n_shots)
+        )
+
+    def KL_whole_training_set(
+        self, target_state: Optional[Array] = None, n_shots: Optional[int] = None
+    ):
+        r"""
+        Compute average KL divergence loss over the whole training set.
+
+        Args:
+            target_state: the target state.
+            n_shots: number of shots per measurement basis.
+
+        .. warn::
+
+            Exponentially expensive in the hilbert space size!
+
+            Need to know the target state!
+
+        """
+        return statistics(
+            self.nll_whole_training_set(return_stats=False)
+            - self.entropy(target_state, n_shots)
+        )
+
+    def KL_exact(
+        self, target_state: Optional[Array] = None, n_shots: Optional[int] = 1
+    ) -> float:
+        r"""
+        Compute the average KL divergence loss between the variational state and the target state.
+
+        Args:
+            target_state: the target state.
+            n_shots: number of shots per measurement basis.
+
+        .. warn::
+
+            Exponentially expensive in the hilbert space size!
+
+            Need to know the target state!
+
+        """
+        if len(target_state.shape) == 1:
+            target_state = target_state.reshape(-1, 1)
+            target_state = target_state @ target_state.conj().T
+        rotations = self._training_rotations[::n_shots]
+        KL_list = []
+        try:
+            vs = self.state.to_matrix(normalize=True)
+        except:
+            vs = self.state.to_array(normalize=True).reshape(-1, 1)
+            vs = vs @ vs.conj().T
```

Comment:
Flake complains about the `except` here, which does not specify a specific exception type to catch. In any case, I think this logic might be better realized without replying on exception handling. A simple alternative could be to check the type of `self.state` via `isinstance` in order to only call `to_matrix` on a type that supports it.

Reply:
Thanks for pointing these out. I did use black, but it seems that these problems were not reported. I've fixed the formatting and removed the try-except.